platform/solr/solr-factory/src/test/groovy/org/codice/solr/factory/impl/SolrClientAdapterAsyncSpec.groovy

```diff
@@ -0,0 +1,1253 @@
+  static final String COLLECTION = 'collection'
+  static final int TIMEOUT_IN_SECS = 25
+  static final String PING_ERROR_MSG = 'failing ping'
+
+  def client = Stub(SolrClient)
+  def creator = Stub(Creator)
+
+  def request = Stub(SolrRequest)
+  def response = Stub(NamedList)
+
+  def pingFail = new UnavailableSolrException("ping failed with 'FAIL'")
+  def pingError = new SolrException(ErrorCode.UNKNOWN, PING_ERROR_MSG)
+
+  def pingOkResponse = Mock(SolrPingResponse) {
+    getResponse() >> Mock(NamedList) {
+      get('status') >> 'OK'
+    }
+  }
+  def pingFailResponse = Mock(SolrPingResponse) {
+    getResponse() >> Mock(NamedList) {
+      get('status') >> 'FAIL'
+    }
+  }
+  def pingErrorResponse = Mock(SolrPingResponse) {
+    getResponse() >> Mock(NamedList) {
+      get('status') >> {
+        throw pingError
+      }
+    }
+  }
+
+  def failsafeCreator = {
+    Mock(SyncFailsafe) {
+      with(_ as ScheduledExecutorService) >> Mock(AsyncFailsafe) {
+        onRetry(_) >> { (delegate as MockInvocation).mockObject.instance } // itself
+        onAbort(_) >> { (delegate as MockInvocation).mockObject.instance } // itself
+        onFailure(_) >> { (delegate as MockInvocation).mockObject.instance } // itself
+        onSuccess(_) >> { (delegate as MockInvocation).mockObject.instance } // itself
+        get(_) >> Stub(FailsafeFuture)
+        run(_) >> Stub(FailsafeFuture)
+      }
+    }
+  }
+
+  @Unroll
+  def 'test constructor when client is created after #created_after and becomes available'() {
+    given: "a client and creator to verify that failsafe actually delegates to our creator and client"
+      def client = Mock(SolrClient) {
+        1 * ping() >> pingOkResponse
+      }
+      def creator = Mock(Creator) {
+        1 * create() >> client
+      }
+
+    and: "controllers that will actually proceed through Solr"
+      def createController = new FailsafeController('SolrClient Creation')
+      def pingController = new FailsafeController('SolrClient Ping').onNextExecution(doProceed())
+
+      if (return_or_throw == SolrClient) {
+        createController.onNextExecution(doProceed())
+      } else {
+        // null or exception
+        createController.onNextExecution {
+          doThrowOrReturn(return_or_throw)
+              .then().doProceed()
+        }
+      }
+
+    when: "creating an adapter"
+      def adapter = new SolrClientAdapter(CORE, creator, createController.&with, pingController.&with)
+
+    then: "the reported core should correspond to the requested one"
+      adapter.core == CORE
+
+    and:
+      "our creator should be registered with the adapter"
+      adapter.creator.is(creator)
+
+    when: "waiting for failsafe to complete"
+      createController.waitForSuccessfulCompletion()
+      pingController.waitForSuccessfulCompletion()
+
+    then: "the adapter should be available"
+      adapter.isAvailable()
+
+    and: "the underlying client should be the one we created"
```

Comment:
We could. But my issue is that since these are more integration-type test cases as they test the behavior of the classes under certain asynchronous stimulus, although that first test case would fail, that second one would continue into other parts which might fail for weird reasons and require more thinking as to why it is failing. Ensuring the little things before proceeding just ensures sure we are in a given state before testing the next part. Thus if there is an issue creating the adapter, all tests would fail at the same place and reduce the amount of investigation that would have to be done.